exp/internal/controllers/machinepool_controller_test.go

```diff
@@ -336,82 +342,141 @@ func TestReconcileMachinePoolRequest(t *testing.T) {
+				mpExist: true,
+				result:  reconcile.Result{},
+				err:     false,
 			},
+			withTracker: true,
 		},
 		{
+			name: "Delete machinePool with nodeDeleteTimeout passed",
 			machinePool: expv1.MachinePool{
 				ObjectMeta: metav1.ObjectMeta{
-					Name:      "deleted",
+					Name:      "deleted-timeoutNode",
 					Namespace: metav1.NamespaceDefault,
 					Labels: map[string]string{
 						clusterv1.MachineControlPlaneLabel: "",
 					},
 					Finalizers:        []string{expv1.MachinePoolFinalizer},
-					DeletionTimestamp: &time,
+					CreationTimestamp: metav1.Time{Time: timeNow.Add(time.Minute * -2)},
+					DeletionTimestamp: &metav1.Time{Time: timeNow.Add(time.Minute * -1)},
 				},
 				Spec: expv1.MachinePoolSpec{
 					ClusterName: "test-cluster",
 					Replicas:    ptr.To[int32](1),
 					Template: clusterv1.MachineTemplateSpec{
 						Spec: clusterv1.MachineSpec{
-							InfrastructureRef: corev1.ObjectReference{
-								APIVersion: builder.InfrastructureGroupVersion.String(),
-								Kind:       builder.TestInfrastructureMachineTemplateKind,
-								Name:       "infra-config1",
-							},
-							Bootstrap: clusterv1.Bootstrap{DataSecretName: ptr.To("data")},
+							NodeDeletionTimeout: &metav1.Duration{Duration: 10 * time.Second},
+						},
+					},
+					ProviderIDList: []string{"aws:///us-test-2a/i-013ab00756982217f"},
+				},
+				Status: expv1.MachinePoolStatus{
+					NodeRefs: []corev1.ObjectReference{
+						{
+							APIVersion: "v1",
+							Kind:       "Node",
+							Name:       "test-node",
 						},
 					},
 				},
 			},
 			expected: expected{
-				result: reconcile.Result{},
-				err:    false,
+				mpExist: false,
+				result:  reconcile.Result{},
+				err:     true,
+			},
+		},
+		{
+			name: "Try delete machinePool with nodeDeleteTimeout set to zero",
+			machinePool: expv1.MachinePool{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "try-delete",
+					Namespace: metav1.NamespaceDefault,
+					Labels: map[string]string{
+						clusterv1.MachineControlPlaneLabel: "",
+					},
+					Finalizers:        []string{expv1.MachinePoolFinalizer},
+					CreationTimestamp: metav1.Time{Time: timeNow.Add(time.Minute * -2)},
+					DeletionTimestamp: &metav1.Time{Time: timeNow.Add(time.Minute * -1)},
+				},
+				Spec: expv1.MachinePoolSpec{
+					ClusterName: "test-cluster",
+					Replicas:    ptr.To[int32](1),
+					Template: clusterv1.MachineTemplateSpec{
+						Spec: clusterv1.MachineSpec{
+							NodeDeletionTimeout: &metav1.Duration{Duration: 0 * time.Second},
+						},
+					},
+					ProviderIDList: []string{"aws:///us-test-2a/i-013ab00756982217f"},
+				},
+				Status: expv1.MachinePoolStatus{
+					NodeRefs: []corev1.ObjectReference{
+						{
+							APIVersion: "v1",
+							Kind:       "Node",
+							Name:       "test-node",
+						},
+					},
+				},
+			},
+			expected: expected{
+				mpExist: true,
+				result:  reconcile.Result{},
+				err:     true,
 			},
 		},
 	}
 
 	for i := range testCases {
```

Comment:
If I understand correctly this is a bit racy:
* In some cases we will only get errors if the patchHelper fails to update the MachinePool, if the MachinePool goes away quickly enough after reconcileDelete removes the finalizer
  * => I would suggest to drop the error check
* In some cases mpExist is set to true, even though we expect the MachinePool to go away after the finalizer is removed
  * => I would recommend to change the check for mpExists to use Eventually & Consistently

```go
			if tc.expected.mpExist {
				g.Consistently(func(g Gomega) {
					g.Expect(r.Client.Get(ctx, key, &expv1.MachinePool{})).To(Succeed())
				}).WithTimeout(2 * time.Second)
			} else {
				// We expect to fail retrieve the machinePool as it is deleted
				g.Eventually(func(g Gomega) {
					g.Expect(apierrors.IsNotFound(r.Client.Get(ctx, key, &expv1.MachinePool{}))).To(BeTrue())
				}).WithTimeout(5 * time.Second)
			}
```

Reply:
Once these changes are made, the assertions are correct, but I think the last test case doesn't succeed anymore. Looks like the MP deletion is not actually blocked with our current test setup

Reply:
Not sure if this is still a problem since we merged https://github.com/kubernetes-sigs/cluster-api/pull/10787

Reply:
oh yes, you are right I re-based the PR.